### PR TITLE
Reactor#6 Testing

### PR DESCRIPTION
**Pull request by leonplondon**
# Summary
Added test for implemented features.

# Trello Card
https://trello.com/c/GILEvXm1/12-testing

### CODE REVIEW

File: src/test/java/wolox/reactortraining/controllers/BotControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+            )
+            .accept(MediaType.APPLICATION_STREAM_JSON)
+            .exchange()
+            .expectStatus()
+            .isBadRequest();
+    }
+
+    @Test
+    public void givenValidInput_whenConversationIsCalled_thenResponseMustBeSuccess() {
+        String conversation1 = "conversation_1";
```
Comment (Jhovannycanaspino) [inline src/test/java/wolox/reactortraining/controllers/BotControllerTest.java]: delete this variable, it is used as parameter in the Array.aslist

File: src/test/java/wolox/reactortraining/controllers/BotControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+        List<String> conversations = Arrays
+            .asList(conversation1, "conversation_2", "conversation_3", "conversation_4");
```
Comment (Jhovannycanaspino) [inline src/test/java/wolox/reactortraining/controllers/BotControllerTest.java]: you can remove this variable, and call the array directly

File: src/test/java/wolox/reactortraining/controllers/TwitterControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+                .path(searchTweeterUrl)
+                .queryParam("query", query)
+                .build()
+            )
+            .exchange()
+            .expectStatus().isOk()
+            .expectBody()
+            .json(objectMapper.writeValueAsString(searchResults));
+    }
+}
```
Comment (Jhovannycanaspino) [inline src/test/java/wolox/reactortraining/controllers/TwitterControllerTest.java]: add EOF

File: src/test/java/wolox/reactortraining/controllers/UserControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+            .uri(getUsers)
+            .exchange()
+            .expectStatus().isOk()
+            .expectBody()
+            .jsonPath("$").exists()
+            .jsonPath("$").isArray()
+            .jsonPath("$").isNotEmpty()
+            .jsonPath("$", hasSize(2));
+    }
+}
```
Comment (Jhovannycanaspino) [inline src/test/java/wolox/reactortraining/controllers/UserControllerTest.java]: add EOF

File: src/test/java/wolox/reactortraining/repositories/TopicRepositoryTest.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+                assertThat(next.getDescription()).isEqualTo(topic.getDescription());
+                assertThat(next.getUsers().get(0).getId()).isNotNull();
+                assertThat(next.getUsers().get(0).getId()).isEqualTo(user.getId());
+                assertThat(next.getUsers().get(0)).isEqualTo(user);
+            })
+            .expectComplete()
+            .verify();
+    }
+
+}
```
Comment (Jhovannycanaspino) [inline src/test/java/wolox/reactortraining/repositories/TopicRepositoryTest.java]: add EOF

File: src/test/java/wolox/reactortraining/repositories/UserRepositoryTest.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+            .create(findAllFlux)
+            .expectSubscription()
+            .expectComplete()
+            .verify();
+    }
+
+    @Test
+    public void givenPersistedUsersInDB_whenFindAllIsCalled_thenReturnNonEmptyResultSet() {
+        UserDto dto = new UserDto();
+        dto.setUsername(randomAlphanumeric(6));
```
Comment (Jhovannycanaspino) [inline src/test/java/wolox/reactortraining/repositories/UserRepositoryTest.java]: refactoring, this code is repeated several times in the class

File: src/test/java/wolox/reactortraining/repositories/UserRepositoryTest.java [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+                assertThat(next.getTopics()).isNotEmpty();
+                assertThat(next.getTopics()).hasSize(1);
+                assertThat(next.getTopics().get(0)).isEqualTo(topic);
+                assertThat(next.getTopics().get(0).getId()).isNotNull();
+            })
+            .expectComplete()
+            .verify();
+
+    }
+}
```
Comment (Jhovannycanaspino) [inline src/test/java/wolox/reactortraining/repositories/UserRepositoryTest.java]: add EOF

File: src/test/java/wolox/reactortraining/services/TwitterServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        assertThat(tweetsStreamPipe1).isEqualTo(tweetsStreamPipe2);
+
+        verify(streamingOperations, times(1)).sample(streamListeners.capture());
+
+        List<StreamListener> listeners = streamListeners.getValue();
+
+        assertThat(listeners).hasSize(1);
+        assertThat(listeners.get(0)).isInstanceOf(StreamListenerDelegate.class);
+    }
+}
```
Comment (Jhovannycanaspino) [inline src/test/java/wolox/reactortraining/services/TwitterServiceTest.java]: add EOF

File: src/test/java/wolox/reactortraining/services/BotServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+                assertThat(response.contains(botNames.get(0))
+                    || response.contains(botNames.get(1))
+                    || response.contains(botNames.get(2)))
+                    .isTrue();
+                return true;
+            })
+            .expectComplete()
+            .verify();
+    }
+}
```
Comment (Jhovannycanaspino) [inline src/test/java/wolox/reactortraining/services/BotServiceTest.java:282]: same

File: src/test/java/wolox/reactortraining/controllers/TwitterControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+import org.springframework.social.twitter.api.Tweet;
+import org.springframework.test.context.junit4.SpringRunner;
+import org.springframework.test.web.reactive.server.WebTestClient;
+import wolox.reactortraining.services.TwitterService;
+
+@RunWith(SpringRunner.class)
+@WebFluxTest(TwitterController.class)
+public class TwitterControllerTest {
+
+    private final String searchTweeterUrl = "/twitter";
```
Comment (Jhovannycanaspino) [inline src/test/java/wolox/reactortraining/controllers/TwitterControllerTest.java]: searchTwitterUrl

**Review by Jhovannycanaspino — COMMENTED** [review]
good work, very good test implementations, not clear to me the use of mockwebserver

File: src/test/java/wolox/reactortraining/repositories/UserRepositoryTest.java (hunk 6 above)
Comment (leonplondon) [reply src/test/java/wolox/reactortraining/repositories/UserRepositoryTest.java]: I used the `I` from the F.I.R.S.T principles, that dictates each tests should be isolated from other, so each one in its arrange stage must create everything it needs to run and be repeatable.

https://pivotal.io/application-modernization-recipes/testing/spring-boot-testing-best-practices

**Comment by leonplondon** [comment]
> good work, very good test implementations, not clear to me the use of mockwebserver

`MockWebServer` is a library that allows us to mock external API calls over HTTP, so what we do is to change the urls to a local url where the `MockWebServer` is listening on and enqueue mocked responses each time we need one, with the expected behavior according to the associated test.

File: src/test/java/wolox/reactortraining/repositories/UserRepositoryTest.java (hunk 6 above)
Comment (Jhovannycanaspino) [reply src/test/java/wolox/reactortraining/repositories/UserRepositoryTest.java]: The principle that quotes, is about keeping the test independent, ie should only do one thing and not interfere with the other tests, but repeated code, refers to the discipline refactoring. https://refactoring.com/.

**Review by Jhovannycanaspino — COMMENTED** [review]
only one comentary

File: src/test/java/wolox/reactortraining/repositories/UserRepositoryTest.java (hunk 6 above)
Comment (leonplondon) [reply src/test/java/wolox/reactortraining/repositories/UserRepositoryTest.java]: https://hackernoon.com/test-f-i-r-s-t-65e42f3adc17
Habia entendido otro asunto de este articulo

**Review by Jhovannycanaspino — APPROVED** [review]
good job!

**Review by darwinaraqueo-zz — COMMENTED** [review]
Respecto a los test, a futuro podriamos mejorar mucho integrando Spock, y quizas porque no... liquidbase para bases de datos de test, asi no mockear ciertas cosas, pero en su defecto esta muy bien el PR, hay cosas que ver y estandarizar, como en handling de errores :)  https://stackoverflow.com/questions/45365791/spring-webflux-emit-exception-upon-null-value-in-spring-data-mongodb-reactive-r mira elmanejo de errores de aqui, me gista que no nos llenamos de clases UsernotFoun Topic Not Found y así

**Comment by leonplondon** [comment]
> Respecto a los test, a futuro podriamos mejorar mucho integrando Spock, y quizas porque no... liquidbase para bases de datos de test, asi no mockear ciertas cosas, pero en su defecto esta muy bien el PR, hay cosas que ver y estandarizar, como en handling de errores :) https://stackoverflow.com/questions/45365791/spring-webflux-emit-exception-upon-null-value-in-spring-data-mongodb-reactive-r mira elmanejo de errores de aqui, me gista que no nos llenamos de clases UsernotFoun Topic Not Found y así

Perfect, Maybe you can prepare a short training about Spock and liquidbase for future projects.

**Review by yanzacd — APPROVED** [review]

**Pull request merged**